test/Test/Kupo/Data/HydraSpec.hs

```diff
@@ -0,0 +1,110 @@
+-- This Source Code Form is subject to the terms of the Mozilla Public
+-- License, v. 2.0. If a copy of the MPL was not distributed with this
+-- file, You can obtain one at http://mozilla.org/MPL/2.0/.
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+module Test.Kupo.Data.HydraSpec
+    ( spec
+    ) where
+
+import Kupo.Prelude
+
+import qualified Data.Aeson as Json
+import Data.Aeson.Lens
+    ( _Array
+    , key
+    )
+import qualified Data.Aeson.Types as Json
+import qualified Data.Set as Set
+import Kupo.App.ChainSync.Hydra
+    ( TransactionStore (..)
+    , TransactionStoreException (TransactionNotInStore)
+    , newTransactionStore
+    , pushTx
+    )
+import Kupo.Data.Hydra
+    ( decodeHydraMessage
+    )
+import Kupo.Data.PartialBlock
+    ( PartialTransaction (..)
+    )
+import Test.Hspec
+    ( Spec
+    , context
+    , parallel
+    , shouldBe
+    , shouldThrow
+    )
+import Test.Hspec.QuickCheck
+    ( prop
+    )
+import Test.Kupo.AppSpec
+    ( genPartialTransactions
+    )
+import Test.Kupo.Data.Generators
+    ( genOutputReference
+    )
+import Test.QuickCheck
+    ( Property
+    , counterexample
+    , label
+    , listOf1
+    , shuffle
+    , withMaxSuccess
+    )
+import Test.QuickCheck.Monadic
+    ( assert
+    , monadicIO
+    , monitor
+    , pick
+    , run
+    )
+
+spec :: Spec
+spec = parallel $ do
+
+  context "TransactionStore" $ do
+
+    prop "can retrieve transactions in any order" $ monadicIO $ do
+      TransactionStore{pushTx, popTxById} <- run newTransactionStore
+      txs <- pick $ do
+           txIns <- listOf1 genOutputReference
+           evalStateT genPartialTransactions (Set.fromList txIns)
+      txsWithIds <- forM txs $ \tx@PartialTransaction{id} -> do
+            run $ pushTx tx
+            pure (tx, id)
+      monitor (label $ "Generated list length is " <> show (length txsWithIds))
+      shuffledTxs <- pick $ shuffle txsWithIds
+      pure $
+          forM_ shuffledTxs $ \(tx, txId) -> do
+              tx' <- popTxById txId
+              tx' `shouldBe` tx
+              popTxById txId `shouldThrow` \case
+                  TransactionNotInStore txId' -> txId' == txId
```

Comment:
Small note: this test is 'misplaced' as it's testing stuff from `Kupo.App.ChainSync.Hydra`, not from `Kupo.Data.Hydra`. 

I can move things around eventually, so consider this is just a note for myself 😅

Reply:
Let's have @v0d1ch do it? :P